pennylane/hf/particle_number.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for computing the particle number observable.
+"""
+import autograd.numpy as anp
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.hf.hamiltonian import _generate_qubit_operator, _return_pauli, simplify
+
+
+def particle_number(orbitals):
+    r"""Computes the particle number observable :math:`\hat{N}=\sum_\alpha \hat{n}_\alpha`
+    in the Pauli basis.
+
+    The particle number operator is given by
+
+    .. math::
+
+        \hat{N} = \sum_\alpha \hat{c}_\alpha^\dagger \hat{c}_\alpha,
+
+    where the index :math:`\alpha` runs over the basis of single-particle states
+    :math:`\vert \alpha \rangle`, and the operators :math:`\hat{c}^\dagger` and :math:`\hat{c}` are
+    the particle creation and annihilation operators, respectively.
+
+    Args:
+        orbitals (int): Number of *spin* orbitals. If an active space is defined, this is
+            the number of active spin-orbitals.
+
+    Returns:
+        pennylane.Hamiltonian: the particle number observable
+
+    **Example**
+
+    >>> orbitals = 4
+    >>> print(particle_number(orbitals))
+    (2.0) [I0]
+    + (-0.5) [Z0]
+    + (-0.5) [Z1]
+    + (-0.5) [Z2]
+    + (-0.5) [Z3]
+    """
+
+    if orbitals <= 0:
+        raise ValueError(f"'orbitals' must be greater than 0; got for 'orbitals' {orbitals}")
+
+    r = np.arange(orbitals)
+    table = np.vstack([r, r, np.ones([orbitals])]).T
+
+    coeffs = np.array([])
+    ops = []
+
+    for i in table:
+        coeffs = np.concatenate((coeffs, np.array([i[2]])))
+        ops.append([int(i[0]), int(i[1])])
+
+    return qubit_operator((coeffs, ops))
+
+
+def qubit_operator(o_ferm, cutoff=1.0e-12):
```

Comment:
This function is added in a previous [PR](https://github.com/PennyLaneAI/pennylane/pull/2173.). Copying it here to pass the tests.

Reply:
Probably need to copy the function specific tests as well? Or will this be removed from here before merging?

Reply:
It will be removed from here when the mentioned PR is merged.